File: splitty_server/app/schemas/groups.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel

from app.schemas.expenses import ExpenseSummaryOut


class GroupMemberOut(BaseModel):
    user_id: int
    name: str
    email: str
    joined_at: datetime


class GroupListItemOut(BaseModel):
    id: int
    name: str
    created_by: int
    created_at: datetime
    member_count: int


class GroupDetailOut(BaseModel):
    id: int
    name: str
    created_by: int
    created_at: datetime
    members: list[GroupMemberOut]


class CreateGroupIn(BaseModel):
    name: str


class UpdateGroupIn(BaseModel):
    name: Optional[str] = None


class AddMemberIn(BaseModel):
    user_id: int


class BalanceOut(BaseModel):
    user_id: int
    name: str
    net_balance: int


class CreateSettlementIn(BaseModel):
    paid_by: int
    paid_to: int
    amount: int
    settled_at: Optional[datetime] = None
    notes: Optional[str] = None


class SettlementOut(BaseModel):
    id: int
    group_id: int
    paid_by: int
    paid_to: int
    amount: int
    settled_at: datetime
    notes: Optional[str]


class PaginatedSettlements(BaseModel):
    items: list[SettlementOut]
    total: int
    page: int
    page_size: int


class GroupSummaryOut(BaseModel):
    group: GroupDetailOut
    balances: list[BalanceOut]
    recent_expenses: list[ExpenseSummaryOut]

File: splitty_app/api/groups.ts
import { api } from './client';

// ─── Response types (mirror the OpenAPI schemas) ─────────────────────────────

export interface GroupListItem {
  id: number;
  name: string;
  created_by: number;
  created_at: string;
  member_count: number;
}

export interface GroupMember {
  user_id: number;
  name: string;
  email: string;
  joined_at: string;
}

export interface GroupDetail {
  id: number;
  name: string;
  created_by: number;
  created_at: string;
  members: GroupMember[];
}

export interface ExpenseSummary {
  id: number;
  title: string;
  currency: string;
  total_amount: number;   // minor units
  created_by: number;
  created_at: string;
  type: 'simple' | 'itemised';
}

export interface PaginatedExpenses {
  items: ExpenseSummary[];
  total: number;
  page: number;
  page_size: number;
}

export interface Balance {
  user_id: number;
  name: string;
  net_balance: number;    // minor units, positive = owed to them
}

// ─── API functions ────────────────────────────────────────────────────────────

export function listGroups(): Promise<GroupListItem[]> {
  return api.get<GroupListItem[]>('/api/v1/groups');
}

export function getGroup(groupId: number): Promise<GroupDetail> {
  return api.get<GroupDetail>(`/api/v1/groups/${groupId}`);
}

export function getGroupExpenses(groupId: number, pageSize = 100): Promise<PaginatedExpenses> {
  return api.get<PaginatedExpenses>(`/api/v1/groups/${groupId}/expenses?page_size=${pageSize}`);
}

export function getGroupBalances(groupId: number): Promise<Balance[]> {
  return api.get<Balance[]>(`/api/v1/groups/${groupId}/balances`);
}
